pkg/certificate/reconciler_test.go

```diff
@@ -0,0 +1,216 @@
+package certificate
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	tassert "github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/compute/kube"
+	"github.com/openservicemesh/osm/pkg/constants"
+	fakeConfigClientset "github.com/openservicemesh/osm/pkg/gen/client/config/clientset/versioned/fake"
+	"github.com/openservicemesh/osm/pkg/k8s"
+	"github.com/openservicemesh/osm/pkg/k8s/informers"
+	"github.com/openservicemesh/osm/pkg/messaging"
+	"github.com/openservicemesh/osm/pkg/tests"
+)
+
+func TestCheckAndUpdate(t *testing.T) {
+	a := tassert.New(t)
+
+	mrcClient := &fakeMRCClient{}
+	checkStatus := func(ctx context.Context, mrc *v1alpha2.MeshRootCertificate) (bool, error) {
+		if mrc.Status.State != constants.MRCStatePending {
+			return false, fmt.Errorf("incorrect rotation state")
+		}
+		return (mrc.Status.ComponentStatuses.Bootstrap == constants.MRCComponentStatusValidating &&
+				mrc.Status.ComponentStatuses.Gateway == constants.MRCComponentStatusValidating &&
+				mrc.Status.ComponentStatuses.Sidecar == constants.MRCComponentStatusValidating &&
+				mrc.Status.ComponentStatuses.XDSControlPlane == constants.MRCComponentStatusValidating &&
+				mrc.Status.ComponentStatuses.Webhooks == constants.MRCComponentStatusValidating),
+			nil
+	}
+	updateStatus := func(ctx context.Context, mrc *v1alpha2.MeshRootCertificate) error {
+		mrc.Status.State = constants.MRCStateValidating
+		mrc.Status.Conditions = append(mrc.Status.Conditions, v1alpha2.MeshRootCertificateCondition{
+			Type:   constants.MRCConditionTypeValidatingRollout,
+			Status: "True",
+			Reason: "CertificatePassivelyInUse",
```

Comment:
For `Reasons` we should _definietly_ have those exported as constants somewhere. Our API should have a specific set of Reasons that have very clear meanings documented somewhere